### ISSUE TITLE
[Bug]: `vue-component-meta` does not work in monorepo

### ISSUE THREAD

**Issue by thomasaull** [storybookjs/storybook#26220]
### Describe the bug

I'm using the new storybook `8.0.0-beta.5` but I'm not being able to get inference of more complex props and events provided by `vue-component-meta`. *My guess* is that it has something to do with the typescript config in my project:

```
apps
  storybook
    tsconfig.json
  web
    src
      components
        Button.vue
  tsconfing.json
```

If you want I could provide a repro in case you don't have an idea already how to fix this. It's just a bit of work to rip out all the unrelated parts. Just let me know :)

**Comment by vanessayuenn** [comment]
a repro would certainly be helpful! cc @larsrickert @chakAs3 in case they have some insights already.

**Comment by weikinhuang** [comment]
Hi @vanessayuenn we're also running into this issue with our codebase. I've provided a base repro case for you to check out https://github.com/weikinhuang/storybook-26220-repro

```bash
cd packages/storybook
npm run storybook
```

open http://localhost:6006/

See that `Example/Button` has the autogenerated docs, which was created by `npx storybook init`, while `Core/Button3` which lives outside the `packages` directory does not have the auto docs even though it's 99% copy pasted from the `Button` component.

**Comment by thomasaull** [comment]
@weikinhuang Thanks for providing a repro, I didn‘t get around to do that.

**Comment by eflewis** [comment]
Adding some context to https://github.com/storybookjs/storybook/issues/26220#issuecomment-1995166575 above:

We are working on a fresh installation of storybook 8 in a `packages` directory. Our goal is to have this storybook instance display stories for components from the `ui` directory, as well as from components inside that `packages` directory.

The issue is that components from the `ui` directory are able to be added properly to the storybook, but can never get their jsdoc notation converted by autodocs, and their props aren't fully interpolated into controls. This happens in any case where the component is outside of the packages directory, regardless of whether the stories file is inside packages or somewhere else

![Screen Shot 2024-03-13 at 2 20 18 PM](https://github.com/storybookjs/storybook/assets/3331307/4da5b1ed-1199-4718-90e4-3a012b0edb52)

Side-by-side you can see that both button 2 (left) and button 3 (right) are nearly identical. The stories generated are very different though, with button 2 having a full complement of documented controls, and the story for button 3 having only controls for props which have args defined:

<img width="1085" alt="Screen Shot 2024-03-13 at 2 18 37 PM" src="https://github.com/storybookjs/storybook/assets/3331307/79bb74d2-5e09-4fdf-8fc2-bfc1797fed66">
<img width="1111" alt="Screen Shot 2024-03-13 at 2 18 50 PM" src="https://github.com/storybookjs/storybook/assets/3331307/81193d67-feae-4f53-88aa-a6ee62ead366">

**Comment by seanogdev** [comment]
Were seeing issues in quite a large repo where were running out of memory when enabling the vue-component-meta docgen. Storybook is in the same package though, it just cant cope I guess

**Comment by valentinpalkovic** [comment]
@kasperpeulen and @chakAs3 bringing this to your attention. I guess https://github.com/storybookjs/storybook/issues/26220#issuecomment-1995786533 already described very well the problem and a potential solution.

**Comment by chakAs3** [comment]
> After some tests I am pretty sure that [this line](https://github.com/storybookjs/storybook/blob/b95582abcb3fd51a9bd23c5713807b350ffb785b/code/frameworks/vue3-vite/src/plugins/vue-component-meta.ts#L138C40-L139C1) causes the issue:

It is not the line causes the issue but it is rather the reason why it is happening,  in monorepo accessing files has limitation, we can only access source from the root project ( running storybook project), @valentinpalkovic you have dealt with monorepo limitation.

**Comment by chakAs3** [comment]
> I am not 100% sure whats the best way to fix this. Maybe we should use the paths set in the `.storybook/main.ts` to check if contains `../` and adjust the project root accordingly?

This the only solution i can see too for now.  we can even define a rootProject='../' or  in case there is monorepo or any other workspace path where we have our components project

**Comment by valentinpalkovic** [comment]
Usually we use the `getProjectRoot` function from `@storybook/core-common` to determine the project root, like I did here: https://github.com/storybookjs/addon-webpack5-compiler-babel/blob/main/src/preset.ts#L38

At some point we might support an env variable to manually set the project root.

**Comment by chakAs3** [comment]
> At some point we might support an env variable to manually set the project root.

yes it would be better this way or pass the project root in the framework options

**Comment by vhoyer** [comment]
> After some tests I am pretty sure that [this line](https://github.com/storybookjs/storybook/blob/b95582abcb3fd51a9bd23c5713807b350ffb785b/code/frameworks/vue3-vite/src/plugins/vue-component-meta.ts#L138C40-L139C1) causes the issue:
> 
> The project root for `vue-component-meta` is basically set to the root of the current package so everything above this folder which includes the rest of the monorepo is not analyzed.
> 
> I am not 100% sure whats the best way to fix this. Maybe we should use the paths set in the `.storybook/main.ts` to check if contains `../` and adjust the project root accordingly?

idonno, I'm testing this problem here on my monorepo and I've hard coded this line to point to the corrent project root for my components, and it does stop the "Can't find main source file" error, but the controls are still not working, do you think it's a differnte error still?

**Comment by vhoyer** [comment]
couldn't we set a project root by id? like, to search for the closest `tsconfig` in the `id`?

**Comment by synopticum** [comment]
Is there any workaround yet to make vue-component-meta docgen work in monorepo, specifically in nx/vue?

**Comment by kasperpeulen** [comment]
cc @larsrickert

**Comment by mauritomei** [comment]
> Is there any workaround yet to make vue-component-meta docgen work in monorepo, specifically in nx/vue?

Anyone find a solution?

@chakAs3 @synopticum

**Comment by Justin-Arnold** [comment]
Just to bring more attention to this issue, we are running into the same issue, using Storybook 8, and Turborepo using pnpm

**Comment by larsrickert** [comment]
tl;dr: Here is a working example using proper monorepo setup: https://github.com/larsrickert/storybook-issue-26220

As far as I can read from your comments, you are creating two projects/folders in a monorepo and try to directly import a component from the other folder right? For example inside the `apps/storybook/*` folder you are importing a Vue file from `apps/web/*`?

If this is the case, this is not a "real" monorepo setup and therefore propably not intended to work.

Instead, your setup should look like this:

1. Create a workspace, e.g. using [pnpm workspaces](https://pnpm.io/workspaces)
2. Add the `apps/web` package as dependency to your `apps/storybook/package.json` with the version set to `workspace:^`
3. pnpm will do the work now to link your packages together to create a "real" monorepo with linked packages
4. In your Storybook, now import the components that you need from the web package (using the package name like any other used npm package, NOT using relative paths)
5. The docs should be generated properly

**Comment by fabioconti** [comment]
Hi @larsrickert
is the solution you provided working specifically in case of monorepos managed with pnpm workspaces?
I have the same issue of missing meta extraction for vue3 components, but the monorepo has been created and it's managed with rushjs.

**Comment by AlmarAubel** [comment]
We have the same problem we use NX to manage our mono repo. We have a single storybook project which loads all the storybook files from the different project as described in https://nx.dev/technologies/test-tools/storybook/recipes/one-storybook-for-all.

**Comment by larsrickert** [comment]
@fabioconti @AlmarAubel I am not too familiar with those tools but I'd assume that they also have some way of linking packages inside the monorepo, right?

If I remember correctly, NX uses aliases in the tsconfig to link packages which vue-component-meta should be able to pick up to generate the docs correctly.

A reproduction example would be helpful for this :)

**Comment by AlmarAubel** [comment]
@larsrickert  I think i managed to create a reproduction. You can check it out [here](https://github.com/AlmarAubel/sb-nx-monorepo)


This repository demonstrates this  issue with Storybook in a monorepo setup: 

**Steps to reproduce:**
1. Clone the repo: https://github.com/AlmarAubel/sb-nx-monorepo
2. Run `npm i` to install dependencies.
3. Start Storybook with:
   ```
   npx nx run @sb-nx-monorepo/storybook-host:storybook
   ```

**Note:**  
In VButton.stories.ts, the type `PagePropsAndCustomArgs` shows that VS Code correctly infers the expected types from the `VButton` component, even though Storybook does not auto-populate the args in the UI.

**Additional Test Case**

To rule out issues caused by loading components from another library, the `VButtonSB` component was also added directly to the `storybook-host` project and included in its Storybook stories. This helps confirm that the problem is not related to cross-library component usage.

**Comment by AlmarAubel** [comment]
It turned out the fix was simply to explicitly configure the docgen plugin to use vue-component-meta. (See [storybook docs](https://storybook.js.org/docs/get-started/frameworks/vue3-vite#override-the-default-configuration))
After adding this to our .storybook/main.ts, everything started working perfectly again:

```
const config: StorybookConfig = {
  framework: {
    name: '@storybook/vue3-vite',
    options: {
      docgen: {
        plugin: 'vue-component-meta',       
      },
    },
  },
};
```

This solved all our issues with missing controls, props, events, and slots documentation.

**Comment by dongj0316** [comment]
The same problem

After I modified the following code using pnpm patch, everything worked fine.

<img width="895" height="452" alt="Image" src="https://github.com/user-attachments/assets/54755165-c687-4c5f-adb8-2b1b6b66f710" />

**Comment by valentinpalkovic** [comment]
Vue docgen now supports monorepo TypeScript configurations, including root `tsconfig.json` files that reference subprojects.

Enable the new server-side docgen path in `.storybook/main.ts`:

```ts
export default {
  features: {
    experimentalDocgenServer: true,
  },
};
```

The fix will be released in Storybook 10.6. I'm closing this issue, but please open a new issue with a reproduction if your monorepo configuration is still unsupported.
